### PR TITLE
`pow`: port to structured, fix pow bug with mixed cpu/cuda tensors

### PR DESCRIPTION
**Pull request by bdhirsh**
This PR does two things:
* Ports `pow` to be structured
* Fixes a bug with how pow handles mixed cpu and cuda tensors

**bug fix**
Pow is a binary op, and all binary ops that use TensorIterator are currently written to handle the case when one of the inputs is a CUDA tensor, and the other is a zero-dimensional cpu tensor.

`pow` incidentally only handles one of the two cases: it fails when the CUDA tensor is passed as the exponent, e.g. `at::pow(torch.tensor(2.0, device='cpu'), torch.tensor([2, 2], device='cuda'))`. Porting `pow` to structured happened to change the error that was outputted from a `TORCH_CHECK` in TensorIterator to an `INTERNAL_ASSERT` in loop.cuh, so I ended up trying to fix the error and update the tests. I added more details in a comment on the PR.

**notes on the structured port**
Pow is a little weird, so I wrote down a couple of issues I noticed during the port:
* Multiple independent overloads. `pow` has two overloads that have their own cpu/cuda kernels, meaning one doesn't call the other. I have to update the names of the kernel overloads to make the compiler happy, since the codegen would otherwise try to generate two classes with the same name. `pow` actually has 3 overloads that all have `out` variants, so I ported all 3 to structured- one of them just happens to redispatch one of the others in most cases.
* Name propagation. Is name propagation implemented per operator? Or is expected to work for most/all ops by default. Right now it looks like it happens for TensorIterator ops by default. For ops that don't use TensorIterator, we need to explicitly pass the names through to the `set_output()` call in the meta function. This happened to matter for `pow` because it has 3 overloads, but only two of them directly use TensorIterator. I had to pass names directly to `set_output` in the 3rd overload to make tests happy.
*  Lack of `const Tensor &` in the C++ API. It's a goal to slowly make all `Tensor &` arguments const as part of the structured port, but in this case I needed to explicitly cast constness away because one structured kernel called back into the C++ API, which still has ordinary `Tensor &` arguments. This probably isn't something we'll fix soon, since we have boxing logic that actually relies on the `Tensor &` / `const Tensor &` distinction in some places.

Stack from [ghstack](https://github.com/ezyang/ghstack):
* #54085 fix torch.pow type promotion issue
* #53680 port div to structured
* #53679 port sub to structured, fix sub.Scalar bug
* **#53669 port at::pow to structured, fix pow bug with mixed cpu/cuda tensors**
* #52692 port at::mul to structured

Differential Revision: [D27029821](https://our.internmc.facebook.com/intern/diff/D27029821)

### CODE REVIEW

File: aten/src/ATen/native/Pow.cpp [hunk 1]
```diff
@@ -6,85 +6,128 @@
...
-}
+//Tensor pow(const Tensor& base, Scalar exp) {
+  //auto dtype = at::result_type(base, exp);
+  //Tensor result = at::empty_like(base, base.options().dtype(dtype), MemoryFormat::Preserve);
+  //return native::pow_out(result, base, exp);
+//}
 
 Tensor pow(Scalar base, const Tensor& exp) {
-  auto dtype = at::result_type(base, exp);
-  Tensor result = at::empty_like(exp, exp.options().dtype(dtype), MemoryFormat::Preserve);
```
Comment (bdhirsh) [inline aten/src/ATen/native/Pow.cpp:86]: I'm losing the `MemoryFormat::Preserve` here, since I can't forward that information to the overload call. I think that's fine, since (a) it looks like `empty_like` [defaults](https://github.com/pytorch/pytorch/blob/1c9fc38eb2bc687487471e05ea20cd3e2359f538/aten/src/ATen/native/TensorFactories.cpp#L251) to `Memory_format::Preserve` anyway, and (b) I'm relying on passing tests.

File: aten/src/ATen/native/Pow.cpp [hunk 2]
```diff
@@ -6,85 +6,64 @@
...
+TORCH_META_FUNC2(pow, Tensor_Scalar) (const Tensor& base, const Scalar exp) {
   // Numpy compatibility check:
   TORCH_CHECK(!(isIntegralType(base.scalar_type(), true) &&
               exp.isIntegral(true) && exp.toLong() < 0),
               "Integers to negative integer powers are not allowed.");
 
   auto common_dtype = at::result_type(base, exp);
-  TORCH_CHECK(at::can_cast(common_dtype, result.scalar_type()),
-           "result type ", common_dtype, " can't be cast to the desired output type ",
-           result.scalar_type());
```
Comment (bdhirsh) [inline aten/src/ATen/native/Pow.cpp:32]: So, before the port this check ran before TensorIterator creation. After the port though, it would have to live inside of the `TORCH_IMPL_FUNC` block, since it directly uses the output tensor. I ended up removing it because:
- It didn't appear to get hit anywhere in tests- the tests that used to hit this bit of code now fail with an assert from inside of TensorIterator first.
- It seems preferable to not have any dtype/shape asserts happen in the impl() call. Preferably, they should all happen inside the meta call, and this seems like a case that TensorIterator handles.

File: aten/src/ATen/native/Pow.cpp [hunk 3]
```diff
@@ -6,85 +6,64 @@
...
-  return native::pow_out(base, base, other);
-}
-
-Tensor& pow_(Tensor& base, Scalar alpha) {
-  return native::pow_out(base, base, alpha);
-}
-
-Tensor pow(const Tensor& base, const Tensor& exp) {
-  auto dtype = at::result_type(base, exp);
-  Tensor result = at::empty({0}, base.options().dtype(dtype));
```
Comment (bdhirsh) [inline aten/src/ATen/native/Pow.cpp:74]: So this was pretty awful: the functional version of `at::native::pow` creates an empty output tensor, putting on the same device as **base**. This caused an inconsistency with how `pow` deals with mixed CPU/CUDA device inputs. 

If `base` is a CUDA tensor, and `exp` is a cpu tensor with 0 dimension thenTensorIterator is happy, because `pow` uses `build_binary_op` to construct the TensorIterator, which by default sets the `allow_cpu_scalars` flag. If you flip them though, and make `exp` the CUDA tensor, and `base` the cpu tensor with 0 dimension, then this line will put the output tensor on CPU. TensorIterator will then complain that you have an output tensor on a different device than one of the inputs.

This got picked up by the structured port, because the tensor that we pass to TensorIterator is undefined: `build_binary_op(maybe_get_output(), base, exp)`. So TensorIterator ends up putting it on the "common_device", which in this case would be CUDA.

With the structured kernel port we actually end up with an INTERNAL_ASSERT instead in [Loops.cuh](https://github.com/pytorch/pytorch/blob/6da0b94dd84d16f251b181ac9da0e8715f1fff57/aten/src/ATen/native/cuda/Loops.cuh#L94), because the `pow(Tensor, Tensor)` kernel isn't written to directly deal with scalars. Updating it to use `gpu_kernel_with_scalars` instead of `gpu_kernel` fixes the issue though.

**Comment by ezyang** [comment]
> Lack of const Tensor & in the C++ API. It's a goal to slowly make all Tensor & arguments const as part of the structured port, but in this case I needed to explicitly cast constness away because one structured kernel called back into the C++ API, which still has ordinary Tensor & arguments. This probably isn't something we'll fix soon, since we have boxing logic that actually relies on the Tensor & / const Tensor & distinction in some places.

Hmm, actually, we might be able to fix this in the public C++ API. The idea is to swap public C++ API to accept const& rather than mut& (this is BC, because mut& converts to const&, but then keep the dispatcher API accepting mut& (and do a const cast there)). Since this is C++, something probably will break, but it probably wouldn't be too bad.

File: aten/src/ATen/native/Pow.cpp (hunk 3 above)
Comment (ezyang) [reply aten/src/ATen/native/Pow.cpp:74]: Possibly related? #50121

File: aten/src/ATen/native/Pow.cpp [hunk 4]
```diff
@@ -6,85 +6,65 @@
...
-           "result type ", common_dtype, " can't be cast to the desired output type ",
-           result.scalar_type());
+  build_unary_op(maybe_get_output(), base.to(common_dtype));
+}
+
+TORCH_META_FUNC2(pow, Scalar) (const Scalar base, const Tensor& exp) {
+    // This overload doesn't directly use TensorIterator. It attempts to short-circuit,
+    // but otherwise redispatches to the Tensor_Tensor overload.
+    auto dtype = at::result_type(base, exp);
+    // Need to call the set_output() overload of set_output to forward named tensor information.
```
Comment (ezyang) [inline aten/src/ATen/native/Pow.cpp:30]: Well, of course you need to call `set_output`, how else is the operator gonna know what the output size is! (Difference with functions above is `build_unary_op` implicitly calls `set_output` in its guts)

File: test/test_binary_ufuncs.py [hunk 5]
```diff
@@ -650,7 +650,7 @@ def to_np(value):
             for test_case in test_cases:
                 self.assertRaisesRegex(RuntimeError, err_msg, test_case)
         else:
-            if isinstance(base, torch.Tensor):
+            if isinstance(base, torch.Tensor) and base.dim() > 0:
```
Comment (ezyang) [inline test/test_binary_ufuncs.py:653]: "reduction" of test coverage

**Review by ezyang — APPROVED** [review]

File: aten/src/ATen/native/Pow.cpp (hunk 4 above)
Comment (bdhirsh) [reply aten/src/ATen/native/Pow.cpp:30]: oh yeah, my intention with this comment was calling out that we needed to use the "internal guts" overload of `set_output()` since we want to pass names through. Probably not a very useful comment though

File: test/test_binary_ufuncs.py (hunk 5 above)
Comment (bdhirsh) [reply test/test_binary_ufuncs.py:653]: fair point. I'll update this to more explicitly check for the change I'm adding: you can't call `pow(torch.tensor(2), torch.tensor([2, 2], device="cuda"))` with the in-place variant of `pow`, since the output tensor in that case is the base, which is already on cpu.

File: aten/src/ATen/native/Pow.cpp [hunk 6]
```diff
@@ -6,85 +6,64 @@
...
-    namedinference::propagate_names(result, exp);
+    out.fill_(1);
   } else if (!base.isComplex() && base.toDouble() == 1.0) {
-    resize_output(result, exp.sizes());
-    result.fill_(1);
-    namedinference::propagate_names(result, exp);
+    out.fill_(1);
   } else {
-    native::pow_out(result, c10::scalar_to_tensor(base, exp.device()), exp);
+    at::pow_out(const_cast<Tensor&>(out), c10::scalar_to_tensor(base, exp.device()), exp); // redispatch!
```
Comment (ngimel) [inline aten/src/ATen/native/Pow.cpp:65]: I think `scalar_to_tensor` here (and not `wrapped_scalar_tensor`) is the reason for #50121, Scalar argument gets incorrect preference for type promotion. @bdhirsh can you please change what's the behavior of your port wrt type pomotion cases in #50121, and if using `wrapped_scalar_tensor` here fixes it.

Comment (bdhirsh) [reply aten/src/ATen/native/Pow.cpp:65]: Good catch! I tested that locally and it looks like it fixes the type promotion problem mentioned in that issue, for cpu:
```
>>> torch.pow(2, torch.tensor([17], dtype=torch.uint8), out=torch.tensor([0]))
tensor([0])
>>> torch.pow(2, torch.tensor(17, dtype=torch.uint8), out=torch.tensor(0))
tensor(0)
```

The behavior for cuda looks unchanged, however. In both the 0-dim tensor case and the scalar case, the computation isn't performed with uint8. I just want to confirm- the behavior is still wrong for cuda, correct? That just means that this fix only partially solves the issue.
```
>>> torch.pow(2, torch.tensor([17], dtype=torch.uint8, device='cuda'), out=torch.tensor([0], device='cuda'))
tensor([131072], device='cuda:0')
>>> torch.pow(2, torch.tensor(17, dtype=torch.uint8, device='cuda'), out=torch.tensor(0, device='cuda'))
tensor(131072, device='cuda:0')
```

Comment (ngimel) [reply aten/src/ATen/native/Pow.cpp:65]: Yeah, the behavior is wrong in cuda.

File: aten/src/ATen/native/cuda/PowKernel.cu [hunk 7]
```diff
@@ -101,30 +101,30 @@ static inline __host__ __device__ B complex_pow_(B base, E exp) {
 }
 #endif
 
-void pow_tensor_tensor_kernel(TensorIterator& iter) {
+void pow_tensor_tensor_kernel(TensorIteratorBase& iter) {
   if (isComplexType(iter.dtype())) {
     AT_DISPATCH_COMPLEX_TYPES(iter.dtype(), "pow_cuda", [&]() {
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/PowKernel.cu:106]: The reason for wrong behavior on cuda is probably because dispatch here is on `iter.dtype()` (which is output dtype) and not on `iter.common_dtype()`. cc @mruberry

Comment (ezyang) [reply aten/src/ATen/native/cuda/PowKernel.cu:106]: Sounds like a footgun if I've ever heard one

Comment (ngimel) [reply aten/src/ATen/native/cuda/PowKernel.cu:106]: It does! Historically `out` participated in type promotion, so this was (almost always) fine, but then it stopped, and suddenly it wasn't.

File: test/test_binary_ufuncs.py [hunk 8]
```diff
@@ -654,12 +654,16 @@ def to_np(value):
                 actual = base.pow(exponent)
                 self.assertEqual(actual, expected.to(actual))
                 actual = base.clone()
-                if torch.can_cast(torch.result_type(base, exponent), base.dtype):
+                # When base is a 0-dim cpu tensor and exp is a cuda tensor, we exp `pow` to work but `pow_` to fail, since
+                # `pow` will try to create the output tensor on a cuda device, but `pow_` needs to use the cpu tensor as the output
+                if base.dim() == 0 and base.device.type == 'cpu' and exponent.device.type == 'cuda':
+                    regex = 'Expected all tensors to be on the same device, but found at least two devices, cuda.* and cpu!'
```
Comment (imaginary-person) [inline test/test_binary_ufuncs.py:660]: This string isn't being used to generate a runtime exception.

Comment (imaginary-person) [reply test/test_binary_ufuncs.py:660]: I'll add the following line after this one in another PR:
`self.assertRaisesRegex(RuntimeError, regex, torch.pow, base, exp)`

Comment (imaginary-person) [reply test/test_binary_ufuncs.py:660]: Added `self.assertRaisesRegex(RuntimeError, regex, base.pow_, exponent)` if exponent is also a tensor.

**Comment by ezyang** [comment]
Lol, should have reviewed this better, uncovered a bug in the Scalar-Tensor handling at #55150
